Definitions from specdev, plus one new definition (mojibake)

aphillips opened the pull request:
- fixed the typo on user-perceived character causing other specs not to find it
- Edits of character encoding to add 'character encoding form'
- Created 'extended grapheme cluster' definition from the wording in specdev (but reworded)
- Added a definition for 'universal character set' and Unicode. Perhaps 'Unicode' could be controversial?


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/aphillips/i18n-glossary/pull/7.html" title="Last updated on Jan 24, 2022, 4:54 PM UTC (e7ee85f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/i18n-glossary/7/de2c2d1...aphillips:e7ee85f.html" title="Last updated on Jan 24, 2022, 4:54 PM UTC (e7ee85f)">Diff</a>

aphillips committed:
Additions and edits copied from w3c/bp-i18n-specdev
- Also includes a definition for "mojibake"

aphillips committed:
Fixed typo in user-percEIved character.

r12a commented on index.html:
```diff
@@ -182,6 +183,8 @@ <h2>Glossary</h2>
 
 <p><dfn id="def_code_point" class="lint-ignore">Code point</dfn>. A code point value represents the position of a character in a coded character set. For example, the code point for the letter <span class="qchar">á</span> in the Unicode coded character set is 225 in decimal, or 0xE1 in hexadecimal notation. Hexadecimal notation is commonly used for referring to code points. See also <a class="ter ...
 
+<p><dfn data-lt="code unit|code units" id="def_code_unit" class="lint-ignore">Code unit</dfn>. The units of data used by a <a>character encoding</a>to encode or serialize characters into a programming language or other serialized form (such as a file). Common code units are 8-, 16-, and 32-bits in size. For example, in C, a <code>char</code> is generally an 8-bit byte: each <code>char</code> is a ...
```
I made a conscious decision at one point not to include _code unit_ in the glossary, because frankly i find that it is not actually a concept that is needed much unless you are deeply into coding theory (which i think most of our readers will not be), and that for others having to factor in this terminology just tends to make things more confusing.  Yes, the terms exist, but in what way does it help to define them here? The big exception, of course, is that code units become important when talking about older uses of JavaScript for supplementary characters, and perhaps we could adapt the definition to focus on that?

aphillips replied:
It is difficult to talk about encodings without being drawn into the code unit/code point distinction. Notice that [Infra](https://infra.spec.whatwg.org/#strings) defines and uses code units as a term (where they mean JS chars).

We could focus the explanation on JavaScript (and DOM), since that's the most interesting code unit on the Web.

r12a commented on index.html:
```diff
@@ -168,7 +169,7 @@ <h2>Glossary</h2>
 <p><dfn data-lt="canonical Unicode locale identifier|canonical tag|canonical locale|Canonical Unicode locale identifier" data-plurals="canonical unicode locale identifiers" id="def_canonical_unicode_locale_identifier" class="lint-ignore">Canonical Unicode locale identifier</dfn>. A <a href="#def_well_formed" class="termref">well-formed language tag</a> resulting from the application of the <a hre ...
 
-<p><dfn id="def_encoding" class="lint-ignore">Character encoding</dfn>. The way the <a class="termref" href="#def_coded_character_set">coded character set</a> is mapped to bytes for manipulation in a computer. Commonly referred to as just the <span class="alt_name">encoding</span>. For examples and further descriptions see <a href="https://www.w3.org/International/articles/definitions-characters/ ...
+<p><dfn data-lt="character encoding|character encoding form" id="def_encoding" class="lint-ignore">Character encoding</dfn> or, more formally, a <em>character encoding form</em>. The way a <a class="termref" href="#def_coded_character_set">coded character set</a> is mapped to <a>code units</a> (such as bytes) for manipulation in a computer. Commonly referred to as just the <span class="alt_name"> ...
```
I had intentionally written that CCS maps to bytes, because they all do, and that's much simpler as an explanation. See the comment on code units below.

aphillips replied:
Thirty years of being fastidious about my jargon makes me nervous here. I don't prefer "character encoding form", but various Unicadetti have taken me to task for not using it. Similarly, once I introduced `code unit` it felt ungenerous not to use it in `character encoding`'s definition. I could put the focus back on bytes easily enough. I'll try an edit.

r12a commented on index.html:
```diff
@@ -209,6 +212,8 @@ <h2>Glossary</h2>
 
 <div class="letter_anchor" id="e">E</div>
 
+<p><dfn data-lt="extended grapheme cluster|extended grapheme clusters" id="def_extended_grapheme_cluster" class="lint-ignore">Extended grapheme cluster</dfn>. A specific type of <a>grapheme</a> defined by [[Unicode]] that most closely matches the mapping of the underlying logical character sequence to a <a>user-perceived character</a>. Extended grapheme clusters are needed because the relationshi ...
```
We already have a definition for grapheme cluster in the glossary which talks about EGCs, and i think that the entry for extended GC should not re-explain what a grapheme cluster is, but should just point to the main entry, where it already explains that it is one of _two_ types of grapheme cluster defined by the Unicode Standard, and the one which should be used. The main entry also points to where the reader can find further information.  I think we can do better in linking than to just 'throw a book at' people using [Unicode]. ;-)

aphillips replied:
Will fix. Note that this is (with edits) what is in specdev now... :-)

r12a commented on index.html:
```diff
@@ -317,7 +322,7 @@ <h2>Glossary</h2>
 <p><dfn data-lt="metadata|metadata" id="def_metadata" class="lint-ignore">Metadata</dfn> is additional information about data. Key types of metadata for <a class="termref" href="#internationalization">internationalization</a> are <a class="termref" href="#def_language_metadata">language metadata</a> and metadata to support <a class="termref" href="#def_bidirectional_text">bidirectional text</a>.  ...
 
-
+<p><dfn id="def_mojibake" class="lint-ignore">Mojibake</dfn> (文字化け). Garbled or incorrectly rendered or processed characters, generally caused by using the wrong <a>character encoding</a> to interpret the <a>code units</a> in a string or file. The word is Japanese in origin and is pronounced approximately <q>moh-gee-bah-kay</q>. For example, the word <q>文字化け</q> encoded as UTF-8 might be displaye ...
```
If we're going to provide pronunciations, can we please use IPA.  I can think of a number of ways people could pronounce moh-gee-bah-kay, and if i didn't know better that would actually suggest the wrong encoding to me.  IPA is /mo.d͡ʒi.ba.ke/.

> if viewed using the (incorrect) character encoding

I suggest: if viewed in an application that thinks (incorrectly) that the character encoding is

aphillips replied:
Happy to fix. Knew you'd provide this comment :-D.

r12a commented on index.html:
```diff
@@ -380,12 +385,14 @@ <h2>Glossary</h2>
 
 <p><dfn data-lt="unicode locale|unicode locale identifier|unicode locale identifiers|unicode locales|Unicode Locale Identifier" id="def_unicode_locale" class="lint-ignore">Unicode Locale Identifier</dfn> or <em>Unicode Locale</em>. A <a href="#def_language_tag" class="termref">language tag</a> that follows the additional rules and restrictions on subtag choice defined in UTR#35 [[UAX35]]. Any val ...
 
+<p><dfn data-lt="universal character set|unicode" id="def_universal_character_set" class="lint-ignore">Universal Character Set</dfn> or <em>Unicode</em>. The character set defined by the [[Unicode]] Standard and which includes all of the characters used to encode text, including historical or extinct writing systems as well as modern usage, private use, typesetting symbols, and other things, such ...
```
defined subsets seems to imply that they are designed to be subsets

maybe drop 'defined'

aphillips replied:
This is copied from specdev. Happy to remove "defined"

r12a commented on index.html:
```diff
@@ -380,12 +385,14 @@ <h2>Glossary</h2>
 <p><dfn data-plurals="user-facing identifiers" id="def_user_facing_identifier" class="lint-ignore">User-facing identifiers</dfn> are identifiers defined by or assigned by a user in a <a href="#def_vocabulary" class="termref">vocabulary</a> that is intended to be at least potentially visible to end-users (and thus is <a href="#def_localizable_text" class="termref">localizable text</a>).</p>
 
-<p><span class="alt_name">User-percieved character</span>. See <a class="termref" href="#def_grapheme">grapheme</a>.</p>
+<p><span class="alt_name">User-perceived character</span>. See <a class="termref" href="#def_grapheme">grapheme</a>.</p>
```
thank you

aphillips committed:
Responding to @r12a's comments
- Changed defintion of code unit to focus more on javascript
- Removed 'code unit' from 'character encoding' (just used bytes)
- Converted 'extended grapheme cluster' to point to 'grapheme cluster'
- Fixed pronunciation of mojibake, adopted Richard's suggested edit.
- Removed 'defined' from 'universal character set'
- Rewrapped 'grapheme' and 'grapheme cluster'
- Fixed reference to Unicode from 8.0 to 14.0. Note that I checked that
  the section reference is still correctly numbered.

r12a commented on index.html:
```diff
@@ -317,7 +305,7 @@ <h2>Glossary</h2>
 <p><dfn data-lt="metadata|metadata" id="def_metadata" class="lint-ignore">Metadata</dfn> is additional information about data. Key types of metadata for <a class="termref" href="#internationalization">internationalization</a> are <a class="termref" href="#def_language_metadata">language metadata</a> and metadata to support <a class="termref" href="#def_bidirectional_text">bidirectional text</a>.  ...
 
-
+<p><dfn id="def_mojibake" class="lint-ignore">Mojibake</dfn> (文字化け). Garbled or incorrectly rendered or processed characters, generally caused by using the wrong <a>character encoding</a> to interpret the <a>code units</a> in a string or file. The word is Japanese in origin and is pronounced <q>/mo.d&#x0361;&#x0292;i.ba.ke/</q>. For example, the word <q>&#x6587;&#x5B57;&#x5316;&#x3051;</q> encode ...
```
I'm not sure that 'code units' are particularly relevant here. I'd be ready to merge the PR if you're willing to change code units to bytes in this entry.

aphillips replied:
Grr. Meant to make that change but overlooked it. Fixed.

aphillips committed:
Changed 'code units' to 'bytes' in 'mojibake'

r12a commented:
Thanks!

himorin commented on index.html:
```diff
@@ -317,7 +305,7 @@ <h2>Glossary</h2>
 <p><dfn data-lt="metadata|metadata" id="def_metadata" class="lint-ignore">Metadata</dfn> is additional information about data. Key types of metadata for <a class="termref" href="#internationalization">internationalization</a> are <a class="termref" href="#def_language_metadata">language metadata</a> and metadata to support <a class="termref" href="#def_bidirectional_text">bidirectional text</a>.  ...
 
-
+<p><dfn id="def_mojibake" class="lint-ignore">Mojibake</dfn> (文字化け). Garbled or incorrectly rendered or processed characters, generally caused by using the wrong <a>character encoding</a> to interpret the bytes in a string or file. The word is Japanese in origin and is pronounced <q>/mo.d&#x0361;&#x0292;i.ba.ke/</q>. For example, the word <q>&#x6587;&#x5B57;&#x5316;&#x3051;</q> encoded as UTF-8 m ...
```
- how about to add lang=ja to characters not in English?
- how about to mention on U+FFFD, which is often shown for cases of non-Unicode two byte encoding in UTF-8?
- (just for comment, I believe no well confirmed standard there...) [wikipedia じ](https://ja.wikipedia.org/wiki/%E3%81%98) uses U+291 instead of U+292 (and in some Japanese dictionary, just `/ʒi/` within a word...)

r12a commented:
> how about to add lang=ja to characters not in English?

Argh.  I knew i'd forgotten something.  I'll create a patch for it when adding some 'see also' links.

> (just for comment, I believe no well confirmed standard there...) wikipedia じ uses U+291 instead of U+292 (and in some Japanese dictionary, just /ʒi/ within a word...)

See https://r12a.github.io/scripts/jpan/#yotsugana.  I considered this when providing the example, and decided maybe best to just use the simplest rendering, which is closest to what most gaijin would say.  I wasn't trying to get a pinpoint accurate phonetic rendering, just indicate rough pronunciation with generally recognisable (IPA) letters. I can change it if we think it could be better.

r12a commented:
I added a [commit](https://github.com/w3c/i18n-glossary/commit/59dbe732ea3ccbe57852436721253455f6488d38) that fixes the missing lang tag.  It also adds some cross-references for natural language/localizable text/syntactic content.
